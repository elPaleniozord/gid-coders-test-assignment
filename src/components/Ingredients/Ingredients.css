.ingredients-header {
  text-align: left;
}
.ingredients-group {
  display: flex;
  align-items: center;
  margin: .3rem 0 .3rem 0;
}
.ingredients-icon {
  height: 50px;
  margin: 5px;
}
.ingredients-list {
  list-style: none;
  display: flex;
  flex-wrap: wrap;
  padding: 0;
}
.ingredients-item {
  margin: 2px 3px 2px 3px;
  font-weight: 600;
}