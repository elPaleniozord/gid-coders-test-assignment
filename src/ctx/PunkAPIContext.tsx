import React, { useEffect, useReducer } from "react"
import { ActionType, punkAPIReducer } from "./PunkAPIReducer"

export interface Beer {
  id: number
  [key: string]: string | number | boolean
}

interface APIContext {
  loading: boolean
  beers: Beer[]
  favorites: number[]
  fetchMore?: () => void
  handleFav?: (beer:Beer) => void
  // dispatch: () => null
}

const defaults = {
  loading: false,
  beers: [],
  favorites: []
}

export const PunkAPIContext = React.createContext<APIContext>(defaults)

const initialState = {
  beers: [],
  loading: false,
  page: 1,
  favorites: []
}

export const PunkAPIProvider: React.FC = ({children}) => {
  const [state, dispatch] = useReducer(punkAPIReducer, initialState)

  const handleLocalStorage: ()=> Beer[] = () => {
      const stored = localStorage.getItem('favorite-beers')
      return stored ? JSON.parse(stored) : []
  }

  const handleFav = async (beer: Beer) => {
    const idx = state.favorites.indexOf(beer.id)
    const update = handleLocalStorage()

    if(idx === -1) update.push(beer)
    else update.splice(idx,1)
    
    localStorage.setItem('favorite-beers', JSON.stringify(update))
    const ids = update.map(beer => beer.id)

    dispatch({type: ActionType.UpdateFavorites, payload: ids})
  }

  useEffect(() => {
    const favs = handleLocalStorage().map(beer => beer.id)
    dispatch({type: ActionType.ToggleLoading})
    fetch(`https://api.punkapi.com/v2/beers?page=${state.page}&per_page=${40}`)
      .then(res => {
        if(res.ok) return res.json()
        throw res
      })
      .then(data => {
        dispatch(
          state.page === 1 ? 
          { type: ActionType.SetBeers, payload: [data, favs] }
          :
          { type: ActionType.UpdateBeers, payload: data }
        )
      })
      .catch(err => console.log(err))
  }, [state.page])

  return (
    <PunkAPIContext.Provider 
      value={{
        loading: state.loading,
        beers: state.beers,
        favorites: state.favorites,
        fetchMore: () => dispatch({type: ActionType.FetchNextPage}),
        handleFav: handleFav,
      }}
    >
      {children}
    </PunkAPIContext.Provider>
  )
}