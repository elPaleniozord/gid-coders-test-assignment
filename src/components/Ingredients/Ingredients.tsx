import './Ingredients.css'

interface Ingredient {
  name: string
}

interface IngredientsProps {
  [key: string]: Ingredient[]
}

const Ingredients = (props: IngredientsProps) => {
  const processIngredients = (arr: Ingredient[]) => {
    const names = new Set(arr.map(item => item.name))
    let map: (Element | JSX.Element)[] = []
    names.forEach(e => map.push(<li key={e} className='ingredients-item'>{e}</li>))
    return map
  } 

  return (
    <div className="ingredients-container">
      <h3 className="ingredients-header">Ingredients</h3>
      <div className="ingredients-group">
        <img className='ingredients-icon' src={`${process.env.PUBLIC_URL}/hops-icon.png`} alt='hops-icon' />
        <ul className='ingredients-list'>{processIngredients(props.hops)}</ul>
      </div>

      <div className="ingredients-group">
        <img className='ingredients-icon' src={`${process.env.PUBLIC_URL}/malt-icon.png`} alt='malt-icon' />
        <ul className='ingredients-list'>{processIngredients(props.malts)}</ul>
      </div>

      <div className="ingredients-group">
        <img className='ingredients-icon' src={`${process.env.PUBLIC_URL}/yeast-icon.png`} alt='yeast-icon' />
        <ul className='ingredients-list'><li className='ingredients-item'>{props.yeast}</li></ul>
      </div>
    </div>
  )
}

export default Ingredients