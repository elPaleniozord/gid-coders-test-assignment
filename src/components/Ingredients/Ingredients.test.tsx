import { render } from '@testing-library/react';
import { sampleBeer } from '../../ctx/PunkAPIContext.test';
import Ingredients from "./Ingredients"

test('should render Favorites Component properly', () => {
  const { container } = render(
    <Ingredients 
      hops={sampleBeer.ingredients.hops} 
      malts={sampleBeer.ingredients.malt} 
      yeast={sampleBeer.ingredients.yeast}
    />
  )
  const favorites = container.querySelector('div')
  expect(favorites).toMatchSnapshot()
})