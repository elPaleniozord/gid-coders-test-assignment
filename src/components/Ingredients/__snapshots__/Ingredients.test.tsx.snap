// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Favorites Component properly 1`] = `
<div
  class="ingredients-container"
>
  <h3
    class="ingredients-header"
  >
    Ingredients
  </h3>
  <div
    class="ingredients-group"
  >
    <img
      alt="hops-icon"
      class="ingredients-icon"
      src="/hops-icon.png"
    />
    <ul
      class="ingredients-list"
    >
      <li
        class="ingredients-item"
      >
        Ahtanum
      </li>
      <li
        class="ingredients-item"
      >
        Chinook
      </li>
    </ul>
  </div>
  <div
    class="ingredients-group"
  >
    <img
      alt="malt-icon"
      class="ingredients-icon"
      src="/malt-icon.png"
    />
    <ul
      class="ingredients-list"
    >
      <li
        class="ingredients-item"
      >
        Extra Pale
      </li>
    </ul>
  </div>
  <div
    class="ingredients-group"
  >
    <img
      alt="yeast-icon"
      class="ingredients-icon"
      src="/yeast-icon.png"
    />
    <ul
      class="ingredients-list"
    >
      <li
        class="ingredients-item"
      >
        Wyeast 1056 - American Aleâ„¢
      </li>
    </ul>
  </div>
</div>
`;
